### ISSUE TITLE
A filter returning pug does not get parsed. (Not a bug, but this behavior deserves improvement.

### ISSUE THREAD

**Issue by qm3ster** [pugjs/pug#2364]
Filters should be parsed first, so that a filter outputting pug can be compiled.
For example consider
```javascript
filters: {
  markpug: function (block){
      return require("html2pug")(require("jstransformer-marked").render(block)));
    }
```
with
```pug
h1= title
:markpug
  **Amazing #{title}**
```
One would expect this to compile to a template that substitutes the variable twice: 

```html
<h1>Markpug</h1>
<p>
  <strong>
    Amazing Markpug
  </strong>
</p>
```

but it really only returns:

```html
<h1>Markpug</h1>
p
  strong
    Amazing #{title}
```
Since pug already ignores blocks starting with `<` there shouldn't be a problem.
tl;dr: I am not complaining about filters being processed at compile-time not runtime, I am talking about parsing the output of filters as jade at compile time.

**Comment by ForbesLindesay** [comment]
This is an interesting idea.  I'm not sure if the use case is sufficiently common to merrit building into core, but it could be built as a plugin.

**Comment by qm3ster** [comment]
So just steal the filters passed, identify filters that can use it, and then plug the output back into the the document before pug does anything?
I just felt like that duplicates a lot of what pug already does, which is identifying the filter blocks, de-indenting them, passing them to a filter (which pug also manages the registering of from external modules) and then putting it back in. 
Since, like I mentioned, lines starting with `<` are already ignored, this could be accomplished by just moving all filters to the top of processing.
Ideally though, and with performance impossible with a plugin, all filters would be handled on the parsing stage, with the output being put into "write-through" node, except if `filter.outputFormat == 'pug'`, in which case the block would be entered.

Anyhow, never mind me. :smile:

**Comment by jbsulli** [comment]
I was thinking about this and I feel like a better solution might be a pug filter? So you can do nested filters, why not pass the value back to a pug filter? Wouldn't need to do weird testing to see if filters returned pug, html, or text that way.

So for the example above:
```
h1= title
:pug:markpug
  **Amazing #{title}**
```

**Comment by ForbesLindesay** [comment]
@jbsulli the issue with that implementation is that filters are purely compile time, and return plain text/html.  In order to behave in the way you would probably want, the `:pug` filter would need to be a magic special case.  For this reason I think it should either be a plugin, or have it's own syntax (or both)

**Comment by qm3ster** [comment]
The problem is for the case of pre-compiled templates which still have runtime substitutions.
Pug doesn't generate html from pug, it generates an efficient function that when given a data object for the substitutions and switches outputs the html.

Over half a year later, I still don't see any danger in modifying the pipeline as I said, by assuming all filters return pug rather than ready to pass through text. On the other hand I can also see that this has limited usefulness as some filters would mangle pug.

**Comment by ForbesLindesay** [comment]
I'd be interested in supporting having something like that as an option.  Either via different syntax, or via the filters returning some object indicating that they expect to be treated as dynamic.

It wouldn't necessarily need to be that they return pug.  They could just return JavaScript directly.
